editor/src/clj/editor/outline_view.clj

```diff
@@ -40,159 +49,234 @@
+
+(defn- reset-tree-view-state [old {:keys [active-resource-node selection open-resource-nodes active-node-id->selected-node-id->node-id-paths active-node-id->node-id-path->expanded] :as new}]
+  (let [ret (-> old
+                (update :active-node-id->selected-node-id->node-id-paths
+                        (fn [old-active-node-id->selected-node-id->node-id-paths]
+                          (let [old-active-node-id->selected-node-id->node-id-paths (select-keys old-active-node-id->selected-node-id->node-id-paths open-resource-nodes)]
+                            (coll/merge-with
+                              (fn [old-selected-node-id->node-id-paths new-selected-node-id->node-id-paths]
+                                (persistent!
+                                  (reduce-kv
+                                    (fn [acc new-selected-node-id _]
+                                      (if-let [old-node-id-paths (old-selected-node-id->node-id-paths new-selected-node-id)]
+                                        (assoc! acc new-selected-node-id old-node-id-paths)
+                                        acc))
+                                    (transient new-selected-node-id->node-id-paths)
+                                    new-selected-node-id->node-id-paths)))
+                              old-active-node-id->selected-node-id->node-id-paths
+                              active-node-id->selected-node-id->node-id-paths))))
+                (assoc :selection selection
+                       :active-resource-node active-resource-node
+                       :open-resource-nodes open-resource-nodes))
+        ret (update ret :active-node-id->node-id-path->expanded
+                    (fn [old-active-node-id->node-id-path->expanded]
+                      (let [old-active-node-id->node-id-path->expanded (select-keys old-active-node-id->node-id-path->expanded open-resource-nodes)
+                            auto-expanded (-> ret
+                                              :active-node-id->selected-node-id->node-id-paths
+                                              (get active-resource-node)
+                                              auto-expand)]
+                        (coll/merge-with
+                          (fn [old-node-id-path->expanded new-node-id-path->expanded]
+                            (persistent!
+                              (reduce-kv
+                                (fn [acc new-node-id-path _]
+                                  (assoc! acc new-node-id-path (boolean (or (old-node-id-path->expanded new-node-id-path)
+                                                                            (auto-expanded new-node-id-path)))))
+                                (transient old-node-id-path->expanded)
+                                new-node-id-path->expanded)))
+                          old-active-node-id->node-id-path->expanded
+                          active-node-id->node-id-path->expanded))))]
+    ret))
```

Comment:
Kinda annoying... previously, the expanded state was managed implicitly: there was a separate tree view per active resource type, and changing selection just mutated the expanded state of new selection parents, while previously expanded tree items kept being expanded. Now that it's reactive, I have to manage it myself...

Reply:
Still, it feels better to not rely on that internal state and simply project The Truth onto the view. 👍